StatefulSet AutoDelete blog post (#30596)
kubernetes/website: content/en/blog/_posts/2021-12-15-StatefulSet-PVC-Auto-Deletion.md

---
layout: blog

> title: 'StatefulSet PVC Auto-Deletion'

Reviewer: ```suggestion
title: 'Kubernetes 1.23: StatefulSet PVC Auto-Deletion (alpha)'
```

Before this new feature, when a StatefulSet-managed replica disappears, either because the StatefulSet is reducing its replica count, or because its StatefulSet is deleted, the PVC and its backing volume remains and must be manually deleted. While this behavior is appropriate when the data is critical, in many cases the persistent data in these PVCs is either temporary, or can be reconstructed from another source. In those cases, PVCs and their backing volumes remaining after their StatefulSet or replicas have been deleted are not necessary, incur cost, and require manual cleanup. 

## The new StatefulSet PVC retention policy

> If you enable the alpha feature, a StatefulSet spec includes a PersistentVolumeClaim retention policy. This is used to control if and when PVCs created from a StatefulSet’s volumeClaimTemplate are deleted. This first iteration of the retention policy contains two situations where PVCs may be deleted. The first situation is when the StatefulSet resource is deleted (which implies that all replicas are also deleted). This is controlled by the **whenDeleted** policy. The second situation, controlled by **whenScaled** is when the StatefulSet is scaled down, which removes some but not all of the replicas in a StatefulSet. In both cases the policy can either be Retain, where the corresponding PVCs are not touched, or Delete, which means that PVCs are deleted. The deletion is done with a normal [object deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/), so that, for example, all retention policies for the underlying PV are respected.

Reviewer: ```suggestion
If you enable the alpha feature, a StatefulSet spec includes a PersistentVolumeClaim retention policy.
This is used to control if and when PVCs created from a StatefulSet’s `volumeClaimTemplate` are deleted.
This first iteration of the retention policy contains two situations where PVCs may be deleted.

The first situation is when the StatefulSet resource is deleted (which implies that all replicas
are also deleted). This is controlled by the `whenDeleted` policy. The second situation, controlled
by `whenScaled` is when the StatefulSet is scaled down, which removes some but not all of the
replicas in a StatefulSet. In both cases the policy can either be `Retain`, where the corresponding
PVCs are not touched, or `Delete`, which means that PVCs are deleted. The deletion is done with
a normal [object deletion](/docs/concepts/architecture/garbage-collection/), so that, for example,
all retention policies for the underlying PV are respected.
```

  * **whenDeleted is Retain and whenScaled is Delete.** This is similar to the previous case, when there is little benefit to keeping PVCs for fast reuse during scale-up. An example of a situation where you might use this is an Elasticsearch cluster.. Typically you would scale that workload up and down to match demand, whilst ensuring a minimum number of replicas (for example: 3). When scaling down, data is migrated away from removed replicas and there is no benefit to retaining those PVCs. However, it can be useful to bring the entire Elasticsearch cluster down temporarily for maintenance. If you need to take the Elasticsearch system offline, you can do this by temporarily deleting the StatefulSet, and then bringing the Elasticsearch cluster back by recreating the StatefulSet. The PVCs holding the Elasticsearch data will still exist and the new replicas will automatically use them.

Visit the [documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-policies) to see all the details.

## What’s next?

> Enable the feature and try it out! Enable the StatefulSetAutoDeletePVC feature gate on a cluster, then create a StatefulSet using the new policy. Test it out and tell us what you think!

Reviewer: ```suggestion
Enable the feature and try it out! Enable the `StatefulSetAutoDeletePVC` feature gate on a cluster, then create a StatefulSet using the new policy. Test it out and tell us what you think!
```

I'm very curious to see if this owner reference mechanism works well in practice. For example, we realized there is no mechanism in Kubernetes for knowing who set a reference, so it’s possible that the StatefulSet controller may fight with custom controllers that set their own references. Fortunately, maintaining the existing retention behavior does not involve any new owner references, so default behavior will be compatible.

> Tag any issues you find with sig-apps and assign to Matthew Cary ([mattcary](https://github.com/mattcary) at GitHub).

Reviewer: ```suggestion
Please tag any issues you report with the label `sig/apps` and assign them to Matthew Cary ([@mattcary](https://github.com/mattcary) at GitHub).
```
?

---
layout: blog
title: 'Kubernetes 1.23: StatefulSet PVC Auto-Deletion (alpha)'

> date: 2021-12-15

Reviewer: BTW did the release comms team check this date yet?

Author: Yeah, the release person (Mickey on slack) said I should do what https://github.com/kubernetes/website/pull/30502 did.

Reviewer 2: Hello! K8s 1.23 Comms Lead here. We actually have this blog scheduled for 12/16 (so it comes out the day after the [honor PV reclaim policy fix blog post](https://github.com/kubernetes/website/pull/30556). Updating the publishing date. 

```suggestion
date: 2021-12-16
```

Author: sgtm, thx!

> slug: statefulset-pvc-auto-deletion

Reviewer: (nit) To match how the other release articles are doing it, this should be
```suggestion
slug: kubernetes-1-23-statefulset-pvc-auto-deletion
```

  * **whenDeleted and whenScaled are both Delete.** PVCs are deleted immediately when their replica is no longer needed. Note this does not include when a Pod is deleted and a new version rescheduled, for example when a node is drained and Pods need to migrate elsewhere. The PVC is deleted only when the replica is no longer needed as signified by a scale-down or StatefulSet deletion. This use case is for when data does not need to live beyond the life of its replica. Perhaps the data is easily reconstructable and the cost savings of deleting unused PVCs is more important than quick scale-up, or perhaps that when a new replica is created, any data from a previous replica is not usable and must be reconstructed anyway.

>   * **whenDeleted is Retain and whenScaled is Delete.** This is similar to the previous case, when there is little benefit to keeping PVCs for fast reuse during scale-up. An example of a situation where you might use this is an Elasticsearch cluster.. Typically you would scale that workload up and down to match demand, whilst ensuring a minimum number of replicas (for example: 3). When scaling down, data is migrated away from removed replicas and there is no benefit to retaining those PVCs. However, it can be useful to bring the entire Elasticsearch cluster down temporarily for maintenance. If you need to take the Elasticsearch system offline, you can do this by temporarily deleting the StatefulSet, and then bringing the Elasticsearch cluster back by recreating the StatefulSet. The PVCs holding the Elasticsearch data will still exist and the new replicas will automatically use them.

Reviewer: Within the bold bits, I'd be tempted to put `whenDeleted` and `whenScale` in backticks (like I did there). I think it'd actually be easier to read like that.

Also, a nit (I think it might actually be my typo from a previous suggestion):
```markdown
this is an Elasticsearch cluster.. Typically
```
two periods

Reviewer: This will also be easier on localization teams if you wrap these at 80 to 100 character length lines, rather than one paragraph per line.

Author: both done.

I also put `Retain` and `Delete` in backticks which is consistent what I've done in the text above.
